OCPCLOUD-2998: implement synchronizedAPI

RadekManak opened the pull request:
### Why:
Adds `.status.synchronizedAPI` field to Machine and MachineSet resources to enable reliable migration cancellation when migrations get stuck at `status.authoritativeAPI: Migrating`. Without this field, the system cannot determine which API was the migration source when users revert `spec.authoritativeAPI`, preventing proper rollback to the last known good state. Implements [OCPCLOUD-2998](https://issues.redhat.com/browse/OCPCLOUD-2998).

### What:
- Adds `.status.synchronizedAPI` field (values: `""` | `MachineAPI` | `ClusterAPI`) to track the last successfully synchronized API
- Implements `handleMigrationStatusInitialization()` in migration controllers to bootstrap empty status fields with proper inference logic
- Adds `IsMigrationCancellationRequested()` detection when `spec.authoritativeAPI` matches `status.synchronizedAPI` while `status.authoritativeAPI == Migrating`
- Updates `ApplyMigrationStatus()` helpers to atomically set both `authoritativeAPI` and `synchronizedAPI` during state transitions

### How can it be used:
Administrators can cancel stuck migrations by reverting `spec.authoritativeAPI` back to the previously synchronized state:

```yaml
# Migration stuck in progress
status:
  authoritativeAPI: Migrating
  synchronizedAPI: MachineAPI  # Last good state

# Cancel by reverting spec
spec:
  authoritativeAPI: MachineAPI  # Matches synchronizedAPI

# System detects cancellation and rolls back
status:
  authoritativeAPI: MachineAPI
  synchronizedAPI: MachineAPI
```

The migration controller detects this pattern and transitions back to the synchronized state without requiring manual intervention.

### How did you test it:
Unit tests cover status initialization scenarios (both fields empty, only one empty, mid-migration inference), migration cancellation detection logic, and rollback flows.

Adds e2e tests to verify field behavior during migrations.

### Notes for the reviewer:
Requires companion PR openshift/machine-api-operator#1442 for API definition vendoring. This PR description was generated with AI assistance.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added API synchronization state tracking to monitor which authority (MachineAPI or ClusterAPI) a resource is synchronized with during migration operations.
  * Implemented migration cancellation logic to automatically revert in-progress migrations when authorities align, improving reliability.

* **Tests**
  * Expanded end-to-end test coverage to verify API synchronization state across machine and machineset migration scenarios.
  * Enhanced unit tests to validate synchronization status during authority transitions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RadekManak committed:
Vendor synchronizedAPI

RadekManak commented:
/pipeline required

RadekManak commented:
go: github.com/openshift/api@v0.0.0-20260105114749-aae5635a71a7 (replaced by ../api): reading ../api/go.mod: open /go/src/github.com/openshift/api/go.mod: no such file or directory

Yes, I switched to using a local project folders for development instead of pushing into my fork. I'll correct this when the API PR is merged.

mdbooth reviewed (commented):
We just discussed this. Unfortunately I don't think this approach is correct.

The purpose of the `SynchronizedAPI` field is to eliminate the heuristic in `isSynchronized` which would fail if we were reverting an incomplete migration. Specifically, the `SynchronizedAPI` field defines deterministically exactly which object the value of `SynchronizedGeneration` refers to. Therefore `SynchronizedAPI` should be written by the sync controller whenever we write `SynchronizedGeneration`.

Additionally, we should ensure that each field is owned by a particular controller and is never touched by any other controller. Therefore the migration controller should not write to this value even to initialise it. It should just wait for the sync controller to initialise it if it is required.

With these changes in place, the additional code path in the migration controller guarded by `IsMigrationCancellationRequested` is not required. As long as we can unambiguously determine the synchronisation state in all circumstances, the existing logic is already correct. Therefore we should just need to update the logic in `isSynchronized` to use the `SynchronizedAPI` field and therefore no longer require the invalid heuristic.

RadekManak committed:
Add SynchronizedAPI migration status to enable rollback

RadekManak commented:
/testwith openshift/cluster-capi-operator/main/e2e-aws-capi-techpreview https://github.com/openshift/machine-api-operator/pull/1442

RadekManak committed:
Fix Makefile to always rebuild binaries

Move FORCE prerequisite before the pipe (|) to make it a regular
prerequisite instead of order-only. This ensures binaries are always
rebuilt when running 'make build', allowing Go's build system to handle
incremental compilation based on source changes.

RadekManak committed:
DNM: Vendor testutils

RadekManak committed:
DNM: bump client-go

RadekManak commented on e2e/machineset_migration_disruptive_test.go:
```diff
@@ -0,0 +1,245 @@
+		fixtureB                machineSetMigrationDisruptiveFixture
+	)
+
+	BeforeAll(func() {
+		if platform != configv1.AWSPlatformType {
+			Skip(fmt.Sprintf("Skipping tests on %s, this is only supported on AWS", platform))
+		}
+
+		if !capiframework.IsFeatureGateEnabled(ctx, cl, features.FeatureGateMachineAPIMigration) {
+			Skip("Skipping, this feature is only supported on MachineAPIMigration enabled clusters")
+		}
+
```
This is N/A. This controller is only deployed on standalone openshift clusters.
Where it this guideline defined?

RadekManak replied:
I see.
https://www.youtube.com/watch?v=eWDrfcxWsuM

stefanonardo commented on e2e/machine_migration_helpers.go:
```diff
@@ -478,3 +478,12 @@ func verifyMachineSynchronizedGeneration(mapiMachine *mapiv1beta1.Machine, autho
 		Fail(fmt.Sprintf("unknown authoritativeAPI type: %v", authority))
 	}
 }
+
+// verifyMachineSynchronizedAPI verifies that the MAPI Machine's status.synchronizedAPI matches the expected value.
+func verifyMachineSynchronizedAPI(mapiMachine *mapiv1beta1.Machine, expectedSynchronizedAPI mapiv1beta1.SynchronizedAPI) {
+	By(fmt.Sprintf("Verifying MAPI Machine SynchronizedAPI is %s", expectedSynchronizedAPI))
```
```suggestion
func verifyMachineSynchronizedAPI(mapiMachine *mapiv1beta1.Machine, expectedSynchronizedAPI mapiv1beta1.SynchronizedAPI) {
    GinkgoHelper()

	By(fmt.Sprintf("Verifying MAPI Machine SynchronizedAPI is %s", expectedSynchronizedAPI))
```

stefanonardo commented on e2e/machineset_migration_helpers.go:
```diff
@@ -289,6 +484,15 @@ func verifyMAPIMachineSetSynchronizedCondition(mapiMachineSet *mapiv1beta1.Machi
 	}, capiframework.WaitMedium, capiframework.RetryMedium).Should(Succeed())
 }
 
+// verifyMachineSetSynchronizedAPI verifies that the MAPI MachineSet's status.synchronizedAPI matches the expected value.
+func verifyMachineSetSynchronizedAPI(mapiMachineSet *mapiv1beta1.MachineSet, expectedSynchronizedAPI mapiv1beta1.SynchronizedAPI) {
+	By(fmt.Sprintf("Verifying MAPI MachineSet SynchronizedAPI is %s", expectedSynchronizedAPI))
```
```suggestion
func verifyMachineSetSynchronizedAPI(mapiMachineSet *mapiv1beta1.MachineSet, expectedSynchronizedAPI mapiv1beta1.SynchronizedAPI) {
    GinkgoHelper()

	By(fmt.Sprintf("Verifying MAPI MachineSet SynchronizedAPI is %s", expectedSynchronizedAPI))
```

stefanonardo commented on pkg/controllers/synccommon/syncstatus.go:
```diff
@@ -124,22 +135,124 @@ func newSyncStatusApplyConfiguration[
+// On success, currentAuthority is guaranteed to be either MachineAPI or ClusterAPI.
+// Missing or invalid SynchronizedAPI values are returned as errors instead of
+// surfacing as a Migrating or otherwise unsupported current authority.
+func MigrationDirection(statusAuthority mapiv1beta1.MachineAuthority, synchronizedAPI mapiv1beta1.SynchronizedAPI, specAuthority mapiv1beta1.MachineAuthority) (mapiv1beta1.MachineAuthority, mapiv1beta1.MachineAuthority, bool, error) {
```
nit for making it more readable for callers:

```suggestion
func MigrationDirection(statusAuthority mapiv1beta1.MachineAuthority, synchronizedAPI mapiv1beta1.SynchronizedAPI, specAuthority mapiv1beta1.MachineAuthority) (currentAuthority, desiredAuthority mapiv1beta1.MachineAuthority, isMigrating bool, err error) {
```
